module/zfs/zfs_vfsops.c

```diff
@@ -1769,6 +1769,13 @@ zfs_remount(struct super_block *sb, int *flags, zfs_mnt_t *zm)
 	if (error)
 		return (error);
 
+	if ((dmu_objset_is_snapshot(zfsvfs->z_os) ||
+	    !spa_writeable(dmu_objset_spa(zfsvfs->z_os))) &&
+	    (!vfsp->vfs_readonly || !(*flags & MS_RDONLY))) {
```

Comment:
I agree that MS_RDONLY is the only authoritative flag for readonly _to the VFS_, but what happens if we have `flags=MS_RDONLY` and `data="rw")`?

I think we should stop processing is this other case too, otherwise we will execute a lot of code (`zfs_register_callbacks/zfs_unregister_callbacks/zfsvfs_vfs_free`) that is useless and potentially dangerous: for instance the panic in `list_destroy()` seems to be caused by unregistered callbacks.